pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompactmerge/UpsertCompactMergeTaskExecutor.java

```diff
@@ -0,0 +1,166 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(UpsertCompactMergeTaskExecutor.class);
+
+  public UpsertCompactMergeTaskExecutor(MinionConf minionConf) {
+    super(minionConf);
+  }
+
+  @Override
+  protected List<SegmentConversionResult> convert(PinotTaskConfig pinotTaskConfig, List<File> segmentDirs,
+      File workingDir)
+      throws Exception {
+    int numInputSegments = segmentDirs.size();
+    List<SegmentConversionResult> results = new ArrayList<>();
+    _eventObserver.notifyProgress(pinotTaskConfig, "Converting segments: " + numInputSegments);
+    String taskType = pinotTaskConfig.getTaskType();
+    Map<String, String> configs = pinotTaskConfig.getConfigs();
+    LOGGER.info("Starting task: {} with configs: {}", taskType, configs);
+    long startMillis = System.currentTimeMillis();
+
+    String tableNameWithType = configs.get(MinionConstants.TABLE_NAME_KEY);
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    Schema schema = getSchema(tableNameWithType);
+
+    SegmentProcessorConfig.Builder segmentProcessorConfigBuilder =
+        new SegmentProcessorConfig.Builder().setTableConfig(tableConfig).setSchema(schema);
+
+    // Progress observer
+    segmentProcessorConfigBuilder.setProgressObserver(p -> _eventObserver.notifyProgress(_pinotTaskConfig, p));
+
+    List<RecordReader> recordReaders = new ArrayList<>(numInputSegments);
+    int count = 1;
+    int partitionId = -1;
+    long maxCreationTimeOfMergingSegments = 0;
+    List<String> originalSegmentCrcFromTaskGenerator =
+        List.of(configs.get(MinionConstants.ORIGINAL_SEGMENT_CRC_KEY).split(","));
+    for (int i = 0; i < numInputSegments; i++) {
+      File segmentDir = segmentDirs.get(i);
+      _eventObserver.notifyProgress(_pinotTaskConfig,
+          String.format("Creating RecordReader for: %s (%d out of %d)", segmentDir, count++, numInputSegments));
+
+      SegmentMetadataImpl segmentMetadata = new SegmentMetadataImpl(segmentDir);
+      String segmentName = segmentMetadata.getName();
+      Integer segmentPartitionId = SegmentUtils.getPartitionIdFromRealtimeSegmentName(segmentName);
+      if (segmentPartitionId == null) {
+        throw new IllegalStateException(String.format("Partition id not found for %s", segmentName));
+      }
+      if (partitionId != -1 && partitionId != segmentPartitionId) {
+        throw new IllegalStateException(String.format("Partition id mismatched for %s, expected partition id: %d",
+            segmentName, partitionId));
+      }
+      partitionId = segmentPartitionId;
+      maxCreationTimeOfMergingSegments = Math.max(maxCreationTimeOfMergingSegments,
+          segmentMetadata.getIndexCreationTime());
+
+      String crcFromDeepStorageSegment = segmentMetadata.getCrc();
+      if (!originalSegmentCrcFromTaskGenerator.get(i).equals(crcFromDeepStorageSegment)) {
+        String message = String.format("Crc mismatched between ZK and deepstore copy of segment: %s. Expected crc "
```

Comment:
Yes one is the deepstore upload retry task where it can upload from any replica and cause CRC mismatch. One more might be an edge case during segment commit time maybe (I have not deep-dived into this). We have an issue to track this - https://github.com/apache/pinot/issues/13491